core/command_parser.h

```diff
@@ -0,0 +1,246 @@
+	void _save_parsed_command(const CommandOption *p_command, int p_idx, int p_args_count);
+
+	// Help text printers
+	String _get_usage(const Vector<Pair<const CommandOption *, String>> &p_printable_commands) const;
+	String _get_commands_description(const OrderedHashMap<String, PackedStringArray> &p_categories_data) const;
+
+	// Other utilies
+	String _to_string(const PackedStringArray &p_names) const; // Returns all command names separated by commas with all prefix variants.
+	String _get_prefixed_longest_name(const PackedStringArray &p_names) const; // Returns longest name with first available prefix (long or short).
+	ParsedPrefix _parse_prefix(const String &p_arg) const;
+	String _find_most_similar(const String &p_name) const;
+	static bool _contains_optional_commands(const Vector<Pair<const CommandOption *, String>> &p_printable_commands);
+
+protected:
+	static void _bind_methods();
+
+public:
+	Error parse_args(const PackedStringArray &p_args);
+	Error notify();
+
+	void add_command(const Ref<CommandOption> &p_command);
+	int get_commands_count() const;
+	Ref<CommandOption> get_command(int p_idx) const;
+	void set_command(int p_idx, const Ref<CommandOption> &p_command);
+	void remove_command(int p_idx);
+	Ref<CommandOption> find_command(const String &p_name) const;
+
+	Ref<CommandOption> add_help_command();
+	Ref<CommandOption> add_version_command();
+
+	bool is_set(const Ref<CommandOption> &p_command) const;
+	String get_value(const Ref<CommandOption> &p_command) const;
+	PackedStringArray get_values(const Ref<CommandOption> &p_command) const;
+
+	String get_prefix(const Ref<CommandOption> &p_command) const;
+	PackedStringArray get_prefixes(const Ref<CommandOption> &p_command) const;
+
+	int get_count(const Ref<CommandOption> &p_command) const;
+
+	PackedStringArray get_forwarded_args() const;
+	PackedStringArray get_args() const;
+	String get_help_text() const;
+	String get_error() const;
+	void clear();
+
+	void set_long_prefixes(const PackedStringArray &p_prefixes);
+	PackedStringArray get_long_prefixes() const;
+	void set_short_prefixes(const PackedStringArray &p_prefixes);
+	PackedStringArray get_short_prefixes() const;
+	void set_help_header(const String &p_header);
+	String get_help_header() const;
+	void set_help_footer(const String &p_footer);
+	String get_help_footer() const;
+	void set_usage_executable(const String &p_name);
+	String get_usage_executable() const;
+	void set_min_command_similarity(double p_similarity);
+	double get_min_command_similarity() const;
+	void set_left_help_pad(int p_size);
+	int get_left_help_pad() const;
+	void set_right_help_pad(int p_size);
+	int get_right_help_pad() const;
+	void set_help_line_length(int p_length);
+	int get_help_line_length() const;
+	void set_min_description_length(int p_length);
+	int get_min_description_length() const;
+	void set_autogenerate_usage(bool p_generate);
+	bool is_autogenerate_usage() const;
```

Comment:
Sounds reasonable. But with dictionary user will not have completion... Maybe introduce a separate class for it?

Reply:
In my opinion, it's not something that you'd like to configure in 90% of the cases, unless you write game-specific parser. But even if you do write game-specific parser, then you might not always be interested in representing the help text as raw text. Padding could also be implemented differently if you don't always use text for that (with `Container` nodes, for instance).

There are indeed classes that don't use `Dictionary`, like https://docs.godotengine.org/en/stable/classes/class_kinematiccollision.html (which *used* to be a dictionary, if I recall correctly), but it was converted to a class because it's used a lot more often in comparison.

So, if I were you I'd just document the formatting options (essentially a list of `Dictionary` keys), like in https://docs.godotengine.org/en/stable/classes/class_physics2ddirectspacestate.html. If that's something that users will configure a lot, then it would make sense to convert formatting options into a class indeed.

Reply:
Did not know about `Physics2DShapeQueryParameters`. But it accepts `Physics2DShapeQueryParameters` derived from a reference, not a `Dictionary`. Is this what you suggest?

Reply:
Looking more closely, I rarely see `Dictionary` used as parameters. I think it would be fine to implement `CommandLineFormatParameters` or `CommandLineFormatOptions` class. But then again, I don't think it's something which would be worth having in the class reference, but maybe it's just me.

Reply:
Hm... I understand the problem, let me think more about it a little more.